crates/lsp-ai/src/memory_backends/vector_store.rs

```diff
@@ -0,0 +1,1062 @@
+        for result in results? {
+            for (sub_result_score, sub_result_chunk) in result {
+                let sub_result_score = if rerank_top_k.is_some() {
+                    match &sub_result_chunk.vec {
+                        StoredChunkVec::Binary(b) => {
+                            // Convert binary vector to f32 vec
+                            let mut b_f32 = vec![];
+                            for byte in b {
+                                for i in 0..8 {
+                                    let x = byte >> (8 - i) & 1;
+                                    b_f32.push(x as f32);
+                                }
+                            }
+                            b_f32.truncate(embedding.len());
+                            #[cfg(feature = "simsimd")]
+                            {
+                                OrderedFloat(
+                                    SpatialSimilarity::dot(&b_f32, &embedding)
+                                        .context("mismatch in vector length when taking the dot product when re-ranking")?
+                                        as f32,
+                                )
+                            }
+                            #[cfg(not(feature = "simsimd"))]
+                            {
+                                OrderedFloat(dot_product(&b_f32, &embedding) as f32)
+                            }
+                        }
+                        StoredChunkVec::F32(_) => {
+                            warn!("Not reranking in vector_store because vectors are not binary");
+                            sub_result_score
+                        }
+                    }
+                } else {
+                    sub_result_score
+                };
+
+                // Filter out chunks that are in the current chunk
+                if sub_result_chunk.uri == current_uri
+                    && sub_result_chunk.range.start_byte <= current_byte
+                    && sub_result_chunk.range.end_byte >= current_byte
+                {
+                    continue;
+                }
+                if top_results.is_empty() {
+                    top_results.insert(sub_result_score, sub_result_chunk);
+                } else if top_results.first_key_value().unwrap().0 < &sub_result_score {
+                    if top_results.len() == limit {
+                        top_results.pop_first();
+                    }
+                    top_results.insert(sub_result_score, sub_result_chunk);
+                }
+            }
+        }
+        Ok(top_results
+            .into_iter()
+            .rev()
+            .map(|(_, chunk)| chunk.text.to_string())
+            .collect())
+    }
+}
+
+pub struct VectorStore {
+    file_store: Arc<FileStore>,
+    crawl: Option<Arc<Mutex<Crawl>>>,
+    splitter: Arc<Box<dyn Splitter + Send + Sync>>,
+    embedding_model: Arc<Box<dyn EmbeddingModel + Send + Sync>>,
+    vector_store: Arc<RwLock<VS>>,
```

Comment:
Have you compared single threaded performance (w/out rayon & the lock) vs lock + rayon?

Reply:
Ah but tokio, so you have to wrap it in a lock